### DISCUSSION TITLE
Will a COSMIC applet happen?

### DISCUSSION THREAD

**Discussion by TC68** [andyholmes/gnome-shell-extension-gsconnect#1703] [Ideas]
I know its early days, but there are many people on Pop!_OS who use the GSconnect gnome extension and will be keen to use an equivalent on the forthcoming COSMIC desktop environment.

Building it is beyond my abilities, but I was wondering if anyone has thought about it?

**Comment by ferdnyc** [comment]
I guess the question is, will COSMIC (I assume you mean the _new_ COSMIC, written in Rust) support extensions at all? Will it support _JavaScript_ extensions?

So far, all of the applets for the new COSMIC seem to be written in Rust. (They're in [this repo](https://github.com/pop-os/cosmic-applets).) If _all_ extensions need to be written in Rust, then not only are there no plans to port GSConnect to it, but there would be no _point_ in doing so. It'd need to be a complete rewrite anyway — you could just as easily port the `kdeconnect` daemon, instead. (_More_ easily, perhaps, since KDE's C++ is probably closer to Rust than GSConnect's GLib-based JavaScript.)

If COSMIC does support GNOME Shell-compatible JavaScript extensions, then assuming GSConnect doesn't just work out of the box, I'd personally be open to making whatever changes are necessary to make GSConnect compatible with COSMIC... depending how extensive they are.

But I can't seem to find any information on the new desktop's extension support/compatibility, even after looking for it.

**Comment by TC68** [comment]
Thanks - that all makes sense. I hope it can be done

**Comment by Mmanish7** [comment]
I installed cosmic desktop and gnome on my opensuse tumbleweed, and installed GSConnect on gnome.

Turns out gsconnect partly works on cosmic in the sense that phone notifications and ring/pinging. Only thing is there's no GSConnect panel plugin. Of course it can't be installed standalone without gnome desktop.

I would love either this or valent to be integrated with cosmic desktop in the future, over kdeconnect, or a fork of the mentioned called "Cosmic Connect"

**Reply by ferdnyc** [reply]
So it sounds like the daemon is working (which uses standard FreeDesktop notifications), but the GNOME Shell component isn't. In which case, there's nothing to integrate it into the desktop.

The question still becomes, "Is COSMIC compatible with GNOME Shell extensions"? If the answer is no (as seems more and more to be the case), then someone needs to step up to write a COSMIC extension/applet, which uses its desktop integration APIs. (Whatever they are, if they even exist.)

If that's going to be delivered as a Flatpak, then the backend could be GSConnect's `daemon.js` running in `gjs`, it could be `kdeconnectd`, it could be one of the other client backends, or it could be a new implementation in Rust.

But it's increasingly sounding like the frontend will need to be COSMIC-specific. Might be more productive to ask around in a COSMIC developer/community forum, rather than here where the audience self-selects for GNOME Shell users.

**Comment by daniellandau** [comment]
I think cosmic is supposed to have applets as flatpaks so the extension part of GSConnect is probably not going to be useful in any way on cosmic. Probably either the GSConnect daemon or Valent (or maybe even kdeconnect itself, I'm not familiar at all with how that's architected) could be integrated with a cosmic applet, but I don't use cosmic and don't have any plans to work on that.
